tests/components/airpatrol/test_coordinator.py

```diff
@@ -0,0 +1,156 @@
+"""Test the AirPatrol data update coordinator."""
+
+from unittest.mock import AsyncMock, patch
+
+from airpatrol.api import AirPatrolAPI, AirPatrolAuthenticationError, AirPatrolError
+import pytest
+
+from homeassistant.components.airpatrol.const import DOMAIN
+from homeassistant.components.airpatrol.coordinator import (
+    SCAN_INTERVAL,
+    AirPatrolDataUpdateCoordinator,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import ConfigEntryAuthFailed, UpdateFailed
+
+from tests.common import MockConfigEntry
+
+
+def make_config_entry(data=None):
+    """Helper to create a MockConfigEntry with custom data."""
+    if data is None:
+        data = {"email": "test@example.com", "password": "pw"}
+    return MockConfigEntry(
+        domain=DOMAIN,
+        entry_id="test_entry_id",
+        data=data,
+        unique_id="uniqueid",
+        title=data["email"],
+    )
+
+
+@pytest.fixture
+def mock_api():
+    """Mock AirPatrol API."""
+    api = AsyncMock(spec=AirPatrolAPI)
+    api.get_data.return_value = [{"unit_id": "unit1"}]
+    api.get_access_token.return_value = "token"
+    api.get_unique_id.return_value = "uniqueid"
+    api.authenticate = AsyncMock()
+    return api
+
+
+@pytest.mark.asyncio
+async def test_update_data_with_stored_token(hass: HomeAssistant, mock_api) -> None:
+    """Test data update with stored access token."""
+    entry = make_config_entry(
+        {"email": "test@example.com", "password": "pw", "access_token": "token"}
+    )
+    entry.add_to_hass(hass)
+    with (
+        patch(
+            "homeassistant.components.airpatrol.coordinator.AirPatrolAPI",
+            return_value=mock_api,
+        ),
+        patch(
+            "homeassistant.components.airpatrol.coordinator.AirPatrolAPI.authenticate",
+            new_callable=AsyncMock,
+            return_value=mock_api,
+        ),
+    ):
+        coordinator = AirPatrolDataUpdateCoordinator(hass, entry)
+        await coordinator._async_setup()
+        result = await coordinator._async_update_data()
+        assert result == {"unit1": {"unit_id": "unit1"}}
+        mock_api.get_data.assert_awaited()
```

Comment:
Setup the integration and move time to make it refresh and then assert the entity state. Not use internals.

Reply:
No, setup the integration.
Don't use the internals of any platform or the coordinator.